src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala

```diff
@@ -29,19 +36,216 @@ case class NoOpFingerprint() {
+
+  /**
+   * Create a Content object from a specified list of leaf files. Any files not listed here will
+   * NOT be part of the returned object.
+   *
+   * @param files List of leaf files.
+   * @return Content object with Directory tree from leaf files.
+   */
+  def fromLeafFiles(files: Seq[FileStatus]): Content = Content(Directory.fromLeafFiles(files))
+}
+
+case class Directory(name: String, files: Seq[FileInfo] = Seq(), subDirs: Seq[Directory] = Seq())
+
+object Directory {
+
+  /**
+   * Create a Directory object from a directory path by recursively listing its leaf files. All
+   * files from the directory tree will be part of the Directory.
+   *
+   * @param path Starting directory path under which the files will be considered part of the
+   *             Directory object.
+   * @param pathFilter Filter for accepting paths. The default filter is picked from spark
+   *                   codebase, which filters out files like _SUCCESS.
+   * @param throwIfNotExists If true, throw FileNotFoundException if path is not found. If set to
+   *                         false, create a blank directory tree with no files.
+   * @return Directory tree starting at root, and containing the files from "path" argument.
+   */
+  def fromDirectory(
+      path: Path,
+      pathFilter: PathFilter = PathUtils.DataPathFilter,
+      throwIfNotExists: Boolean = false): Directory = {
+    val fs = path.getFileSystem(new Configuration)
+    val leafFiles = listLeafFiles(path, pathFilter, throwIfNotExists, fs)
+
+    if (leafFiles.nonEmpty) {
+      fromLeafFiles(leafFiles)
+    } else {
+      val rootPath = getRoot(path)
+      if (rootPath.equals(path)) {
+        createDirectory(path, Array())
+      } else {
+        Directory(rootPath.toString, Seq(), Seq(createDirectory(path, Array())))
+      }
+    }
+  }
+
+  /**
+   * Create a Content object from a specified list of leaf files. Any files not listed here will
+   * NOT be part of the returned object
+   * Pre-requisite: files list should be non-empty.
+   * Pre-requisite: all files must be leaf files.
+   *
+   * @param files List of leaf files.
+   * @return Content object with Directory tree from leaf files.
+   */
+  def fromLeafFiles(files: Seq[FileStatus]): Directory = {
+    require(
+      files.nonEmpty,
+      s"Empty files list found while creating a ${Directory.getClass.getName}.")
+
+    require(
+      files.forall(!_.isDirectory),
+      "All files must be leaf files for creation of Directory.")
+
+    /* from org.apache.spark.sql.execution.datasources.InMemoryFileIndex. */
+    val leafDirToChildrenFiles = files.toArray.groupBy(_.getPath.getParent)
+
+    // Hashmap from directory path to Directory object, used below for quick access from path.
+    val pathToDirectory = HashMap[Path, Directory]()
+
+    for (path <- leafDirToChildrenFiles.keys) {
+      val files = ListBuffer[FileInfo]()
+      files.appendAll(leafDirToChildrenFiles.getOrElse(path, Array()).map(FileInfo(_)))
```

Comment:
```scala
    for ((path, files) <- leafDirToChildrenFiles) {
      val allFiles = ListBuffer[FileInfo]()
      allFiles.appendAll(files.map(FileInfo(_)))
```
Need to replace `files` with `allFiles` below.

Reply:
Also, `dirPath` makes more sense than `path`, since it will always be a directory.

Reply:
thanks @imback82 , added